### PR TITLE
network: implement bridge networking mode

### PR DESCRIPTION
**Pull request by nickethier**
This PR implements the bridge networking mode as well as port forwarding using the new `to` field on the port stanza.

### CODE REVIEW

File: drivers/docker/network.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
+package docker
+
+import (
+	"fmt"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/hashicorp/nomad/plugins/drivers"
+)
+
+const infraContainerImage = "gcr.io/google_containers/pause-amd64:3.0"
```
Comment (jippi) [inline drivers/docker/network.go]: is this configurable somewhere? We don't really want our infra to reach out to foreign container registries at any time - we mirror _everything_ into ECR in the local AWS region

Comment (nickethier) [reply drivers/docker/network.go]: Yeah I am going to get that plumbed up into a client config option.

Comment (nickethier) [reply drivers/docker/network.go]: Yeah I am going to get that plumbed up into a client config option.

File: client/allocrunner/network_hook_linux.go [hunk 2]
```diff
@@ -0,0 +1 @@
+package allocrunner
```
Comment (schmichael) [inline client/allocrunner/network_hook_linux.go]: What's this file for?

File: website/source/docs/drivers/docker.html.md [hunk 3]
```diff
@@ -728,6 +728,10 @@ plugin "docker" {
       `docker.volumes.enabled` set to false, the labels will still be applied to
       the standard binds in the container.
 
+* `infra_image` - This is the docker image to use when creating the parent
```
Comment (schmichael) [inline website/source/docs/drivers/docker.html.md]: Is `infra` the right term for this image? Isn't it more commonly referred to as a `pause` image? Or we could name it after its use: `net_image`. Not a blocker. The name just seems overly generic for the very specific use it seems to serve.

```suggestion
* `infra_image` - This is the Docker image to use when creating the parent
```

File: client/allocrunner/network_hook_linux.go (hunk 2 above)
Comment (nickethier) [reply client/allocrunner/network_hook_linux.go]: I've removed it in a different branch.

File: scheduler/rank.go [hunk 4]
```diff
@@ -224,6 +225,59 @@ OUTER:
 		}
 		preemptor.SetPreemptions(currentPreemptions)
 
+		// Check if we need task group network resource
+		if len(iter.taskGroup.Networks) > 0 {
```
Comment (preetapan) [inline scheduler/rank.go]: Can the taskGroup and each individual task both have network stanzas? 
?"

**Pull request merged**